File: Scheduler.py
"""
MSD P22385
Themed Entertainment Model Display
File    : Scheduler.py
Author  : Harrison Barnes
Date    : 05/03/2022
Purpose : Driver program for the scheduler
"""

# Includes
import datetime
import csv
import Schedule

"""
Function : SchedulerDriver
Inputs   : None
Outputs  : integer code denoting state of schedule

Fetches time, validates the time within the schedule, and returns
a proper validation code.
"""
def SchedulerDriver():
    # Fetch time and date for the current time
    timeOfDay = FetchTime()
    todaysDate = FetchDate()

    # Validate the time and date within the schedule
    return ValidateTime(timeOfDay, todaysDate)


"""
Function : FetchTime
Inputs   : none
Outputs  : Current time of day
"""
def FetchTime():
    return datetime.datetime.now().time()


"""
Function : FetchDate
Inputs   : none
Outputs  : Current date
"""
def FetchDate():
    return datetime.date.today()


"""
Function : ConvertDayOfWeek
Inputs   : dateObj -  a datetime date object
Outputs  : Day of week (0 = mon, 6 = sun)
"""
def ConvertDayOfWeek(dateObj):
    return datetime.date.weekday(dateObj)


"""
Function : PrintTimeOfDay
Inputs   : none
Outputs  : none

#rints time of day
"""
def PrintTimeOfDay():
    print(FetchTime())


"""
Function : PrintDate
Inputs   : none
Outputs  : none

Prints current date
"""
def PrintDate():
    print(FetchDate())


"""
Function : GetTimestamp
Inputs   : none
Outputs  : timestamp - a string representation of the current time & date

Fetches time and dates, creates a string timestamp, and returns it
"""
def GetTimestamp():
    return FetchTime().isoformat() + FetchDate().strftime(" %y-%m-%d\n")


"""
Function : ValidateTime
Inputs   : timeOfDay - current time of day
Outputs  : return code based on validity of time within the ride schedule

Checks if the given time lies within the user-defined schedule. Returns accordingly
Specifically checks if date is valid. If valid, checks if day of week is valid.
If valid, continue to check the time
"""
def ValidateTime(timeOfDay, todaysDate):
    # Extract day of week and current week and year using ISO calendar format
    currentDOW = todaysDate.isocalendar()[2]
    currentWeek = todaysDate.isocalendar()[1]
    currentYear = todaysDate.isocalendar()[0]

    # Extract current date (ISO format) and month
    currentDate = datetime.date.isoformat(todaysDate)
    currentMonth = todaysDate.month

    # Extract hour and minute for evaluation
    currentMin = timeOfDay.minute
    currentHour = timeOfDay.hour

    # If the current day is set as a valid override, all other date-based invalidate checks are skipped.
    # Useful for things like Imagine RIT that fall on weekends, without having to remove the weekend restriction
    validOverrideFlag = False
    for overrideDate in Schedule.VALID_OVERRIDE_DATE_ARRAY:
        if currentDate == overrideDate:  # Override date detected, set flag high to skip date validation
            validOverrideFlag = True

    # Validate the date ONLY if the current date is not an override date
    if not validOverrideFlag:
        # Validate date
        for date in Schedule.INVALID_DATE_ARRAY:
            if currentDate == date:  # If date is invalid, return invalid code
                return Schedule.INVALID_DATE

        # Validate day of week (DOW)
        for day in Schedule.INVALID_DOW_ARRAY:
            if currentDOW == day:  # If DOW is invalid, return invalid code
                return Schedule.INVALID_DOW

        # Validate week number
        for weekYearTuple in Schedule.INVALID_WEEK_ARRAY:
            # If week is invalid, return invalid code
            # Checks year too for insurance that the week is correct
            if currentYear == weekYearTuple[0] and currentWeek == weekYearTuple[1]:
                return Schedule.INVALID_WEEK

        # Validate month number
        for month in Schedule.INVALID_MONTH_ARRAY:
            if currentMonth == month:  # If month is invalid, return invalid code
                return Schedule.INVALID_MONTH

    # Validate time of day (TOD)
    if Schedule.OPEN_HR < currentHour < Schedule.CLOSE_HR:
        # TOD is within valid hours, no need to check minutes
        return Schedule.VALID_TIME
    elif Schedule.OPEN_HR == currentHour:
        # If hour matches opening hour, check minutes
        if Schedule.OPEN_MN <= currentMin:
            # Time is past opening, valid
            return Schedule.VALID_TIME
        else:
            # Time is not at opening, invalid
            return Schedule.INVALID_MINUTE
    elif Schedule.CLOSE_HR == currentHour:
        # If hour matches closing hour, check minutes
        if Schedule.CLOSE_MN >= currentMin:
            # Time is prior to closing, valid
            return Schedule.VALID_TIME
        else:
            # Time is after closing, invalid
            return Schedule.INVALID_MINUTE
    else:
        # Hour is not within correct timeframe (hours)
        return Schedule.INVALID_HOUR


"""
Function : InitializeSchedule
Inputs   : simcode - defaults to false. Used by the schedule simulator to edit the CSV file pathS
Outputs  : return code - returns an error code or success code based upon the CSV reader status

Reads in CSV schedule files and fills in various arrays to store the schedule in Schedule.py
Returns a success or error code
"""
def InitializeSchedule(simcode=False):
    # Sim_Scheduler.py requires a different path than the driver. This handles the simulation code to set the path
    # correctly
    if simcode:  # Adjusted path for simulator
        csvPath = "../Scheduler_CSV/"
    else:  # Default path for driver
        csvPath = "Scheduler_CSV/"

    # Read Invalid_Dates.csv into INVALID_DATE_ARRAY
    with open(csvPath + "Invalid_Dates.csv", newline="") as csvFile:  # Open csv file with invalid dates
        # Initialize the reader and skip column header line (first line)
        reader = csv.reader(csvFile)
        next(reader)

        for row in reader:  # Iterate through all rows with data and store the data into the correct array
            Schedule.INVALID_DATE_ARRAY.append(row[0] + "-" + row[1] + "-" + row[2])
    # Close csv file to free up memory
    csvFile.close()

    # Read Override_Dates.csv into VALID_OVERRIDE_DATE_ARRAY
    with open(csvPath + "Override_Dates.csv", newline="") as csvFile:  # Open csv file with valid override dates
        # Initialize the reader and skip column header line (first line)
        reader = csv.reader(csvFile)
        next(reader)

        for row in reader:  # Iterate through all rows with data and store the data into the correct array
            Schedule.VALID_OVERRIDE_DATE_ARRAY.append(row[0] + "-" + row[1] + "-" + row[2])
    # Close csv file to free up memory
    csvFile.close()

    # Read Invalid_Weeks.csv into INVALID_WEEK_ARRAY
    with open(csvPath + "Invalid_Weeks.csv", newline="") as csvFile:  # Open csv file with invalid weeks
        # Initialize the reader and skip column header line (first line)
        reader = csv.reader(csvFile)
        next(reader)

        for row in reader:  # Iterate through all rows with data and store the data into the correct array
            Schedule.INVALID_WEEK_ARRAY.append((int(row[0]), int(row[1])))
    # Close csv file to free up memory
    csvFile.close()

    # Read Invalid_Months.csv into INVALID_MONTH_ARRAY
    with open(csvPath + "Invalid_Months.csv", newline="") as csvFile:  # Open csv file with invalid months
        # Initialize the reader and skip column header line (first line)
        reader = csv.reader(csvFile)
        next(reader)

        for row in reader:  # Iterate through all rows with data and store the data into the correct array
            Schedule.INVALID_MONTH_ARRAY.append(int(row[0]))
    # Close csv file to free up memory
    csvFile.close()

    # Read Invalid_DOW.csv into INVALID_DOW_ARRAY
    with open(csvPath + "Invalid_DOW.csv", newline="") as csvFile:  # Open csv file with invalid days of week
        # Initialize the reader and skip column header line (first line)
        reader = csv.reader(csvFile)
        next(reader)

        for row in reader:  # Iterate through all rows with data and store the data into the correct array
            Schedule.INVALID_DOW_ARRAY.append(int(row[0]))
    # Close csv file to free up memory
    csvFile.close()

    return Schedule.CSV_SUCCESS

File: Deprecated C Files/Scheduler.c
// MSD P22835
// Themed Entertainment Model Display
// File    : Scheduler.c
// Author  : Harrison Barnes
// Data    : 05/03/2022
// Purpose : Driver program for the scheduler


//----- Includes -----//
#include "Scheduler.h"
#include <time.h>
#include <stdio.h>


//----- Function Declarations -----//


// Function : Scheduler
// Inputs   : none
// Outputs  : integer code denoting state of schedule
//
// Fetches, validates, and returns
int Scheduler(void);


// Function : FetchTime
// Inputs   : none
// Outputs  : timeOfDay - time stored in time_t
//
// Fetches current time based on time.h's library
time_t FetchTime(void);


// Function : PrintTimeOfDay
// Inputs   : timeOfDay - time stored in time_t
// Outputs  : none
//
// Prints the time of day based on the parameter currentTime
// Follows the format:
//          YYYY-MM-DD HR:MIN:SEC
void PrintTimeOfDay(time_t timeOfDay);


// Function : ValidateTime
// Inputs   : timeOfDay - time stored in time_t to be validated
// Outputs  : int - return code designating if the time is within the schedule
//
// Checks if the given time lies within the user-defined schedule. Returns accordingly
int ValidateTime(time_t timeOfDay);


//----- Function Implementations -----//


// Function : Scheduler
// Inputs   : none
// Outputs  : integer code denoting state of schedule
//
// Fetches, validates, and returns
int Scheduler(void) {
    // Var holds time of day fetched from internet
    time_t timeOfDay = FetchTime();

    // Validate time
    int validCode = ValidateTime(timeOfDay);

    return validCode; // Return validation code
}


// Function : FetchTime
// Inputs   : none
// Outputs  : timeOfDay - time stored in time_t
//
// Fetches current time based on time.h's library
time_t FetchTime(void) {
    return time(NULL); // returns the time_t created from time(NULL)
}


// Function : PrintTimeOfDay
// Inputs   : timeOfDay - time stored in time_t
// Outputs  : none
//
// Prints the time of day based on the parameter currentTime
// Follows the format:
//          YYYY-MM-DD HR:MIN:SEC
void PrintTimeOfDay(time_t timeOfDay) {
    // Convert time to usable form
    struct tm currentTime = *localtime(&timeOfDay);

    // Print the time of day, following the specified format
    // Adjusts values based on the struct's zero declarations (ex: year + 1900)
    printf("time: %d-%02d-%02d %02d:%02d:%02d\n", currentTime.tm_year + 1900, currentTime.tm_mon + 1, currentTime.tm_mday, currentTime.tm_hour, currentTime.tm_min, currentTime.tm_sec);
}


// Function : ValidateTime
// Inputs   : timeOfDay - time stored in time_t to be validated
// Outputs  : int - return code designating if the time is within the schedule
//
// Checks if the given time lies within the user-defined schedule. Returns accordingly
// Specifically checks if date is valid. If valid, checks if day of week is valid. 
// If valid, continue to check the time
int ValidateTime(time_t timeOfDay) {
    // Convert time to usable form
    struct tm currentTime = *localtime(&timeOfDay);

    // Extract date, AKA day of year, 0-365 (365 denotes leap year)
    int date = currentTime.tm_yday;

    // Extract day of week, 0-6
    int day = currentTime.tm_wday;

    // Extract hour
    int hour = currentTime.tm_hour;

    // Extract minute
    int min = currentTime.tm_min;

    // Validate dates. Check if date exists in an invalid date array
    for(int dateIdx = 0; dateIdx < sizeof(invalid_dates) / sizeof(int); dateIdx++) {
        // Check every date in invalid array
        if(date == invalid_dates[dateIdx]) {
            // Date matches invalid date in array. Return invalid date code
            return INVALID_DATE;
        }
    }

    // Validate day. Checks if day is not allowed
    for(int dayIdx = 0; dayIdx < sizeof(invalid_days) / sizeof(int); dayIdx++) {
        // Check every date in invalid array
        if(day == invalid_days[dayIdx]) {
            // Date matches invalid day in array. Return invalid day of week code
            return INVALID_DOW;
        }
    }

    // Validate time of day
    if(TOD_BEGIN_HR <= hour && hour <= TOD_CLOSE_HR) {
        // Time is within hours, check minute
        if (TOD_BEGIN_MIN <= min && min <= TOD_CLOSE_MIN) {
            // Time is within minutes, valid time
            return VALID_TIME;
        } else {
            // Time is not within valid operating minutes, but is within hours
            return INVALID_MINUTE;
        }
    } else {
        // Time is not within valid operating hours
        return INVALID_HOUR;
    }
}
